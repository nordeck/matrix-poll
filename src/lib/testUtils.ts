/*
 * Copyright 2022 Nordeck IT + Consulting GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  PowerLevelsStateEvent,
  RoomEvent,
  RoomMemberStateEventContent,
  StateEvent,
} from '@matrix-widget-toolkit/api';
import {
  GroupContent,
  IPoll,
  IPollSettings,
  IVote,
  PollStartEvent,
  PollType,
  ResultType,
  RoomNameEvent,
} from '../model';

/**
 * Create a matrix room member event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockRoomMember({
  state_key = '@user-alice',
  event_id = '$event-id-0',
  content = {},
}: {
  state_key?: string;
  event_id?: string;
  content?: Partial<RoomMemberStateEventContent>;
} = {}): StateEvent<RoomMemberStateEventContent> {
  return {
    type: 'm.room.member',
    sender: '@user-id',
    content: {
      membership: 'join',
      displayname: 'Alice',
      avatar_url: 'mxc://alice.png',
      ...content,
    },
    state_key,
    origin_server_ts: 0,
    event_id,
    room_id: '!room-id',
  };
}

/**
 * Create a matrix room name event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockRoomName({
  room_id = '!room-id',
  content = {},
}: {
  room_id?: string;
  content?: Partial<RoomNameEvent>;
} = {}): StateEvent<RoomNameEvent> {
  return {
    type: 'm.room.name',
    sender: '',
    content: {
      name: 'My Room',
      ...content,
    },
    state_key: room_id,
    origin_server_ts: 0,
    event_id: '$event-id-0',
    room_id,
  };
}

/**
 * Create a matrix power levels event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockPowerLevelsEvent({
  content = {},
}: {
  content?: Partial<PowerLevelsStateEvent>;
} = {}): StateEvent<PowerLevelsStateEvent> {
  return {
    type: 'm.room.power_levels',
    sender: '@user-id',
    content: {
      users_default: 100,
      ...content,
    },
    state_key: '',
    origin_server_ts: 0,
    event_id: '$event-id-0',
    room_id: '!room-id',
  };
}

/**
 * Create a poll event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockPoll({
  state_key = 'poll-0',
  event_id = '$event-id-0',
  origin_server_ts = 0,
  content = {},
}: {
  state_key?: string;
  event_id?: string;
  origin_server_ts?: number;
  content?: Partial<IPoll>;
} = {}): StateEvent<IPoll> {
  return {
    type: 'net.nordeck.poll',
    sender: '@user-id',
    content: {
      title: 'My Title',
      question: 'My Question',
      description: 'My Description',
      pollType: PollType.Open,
      answers: [
        { id: '1', label: 'Yes' },
        { id: '2', label: 'No' },
      ],
      resultType: ResultType.Visible,
      duration: 1,
      groups: [
        {
          id: 'red-party',
          eventId: '$event-id-0',
          abbreviation: 'Red Party',
          color: '#ff0000',
          votingRights: {
            '@user-alice': {
              state: 'active',
            },
            '@user-bob': {
              state: 'active',
            },
            '@user-charlie': {
              state: 'active',
            },
          },
        },
        {
          id: 'blue-party',
          eventId: '$event-id-1',
          abbreviation: 'Blue Party',
          color: '#0000ff',
          votingRights: {
            '@user-dameon': {
              state: 'active',
            },
          },
        },
      ],
      ...content,
    },
    state_key,
    origin_server_ts,
    event_id,
    room_id: '!room-id',
  };
}

/**
 * Create a poll start event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockPollStart({
  event_id = '$start-event-id',
  origin_server_ts = 0,
  content = {},
}: {
  state_key?: string;
  event_id?: string;
  origin_server_ts?: number;
  content?: Partial<PollStartEvent>;
} = {}): RoomEvent<PollStartEvent> {
  return {
    type: 'net.nordeck.poll.start',
    sender: '@user-id',
    content: {
      ...content,
    },
    origin_server_ts,
    event_id,
    room_id: '!room-id',
  };
}

/**
 * Create a group event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockGroup({
  state_key = 'group-0',
  event_id = '$event-id-0',
  content = {},
}: {
  state_key?: string;
  event_id?: string;
  content?: Partial<GroupContent>;
} = {}): StateEvent<GroupContent> {
  return {
    type: 'net.nordeck.poll.group',
    sender: '@user-id',
    content: {
      abbreviation: 'GROUP 0',
      color: '#07f556',
      members: {},
      ...content,
    },
    state_key,
    origin_server_ts: 0,
    event_id,
    room_id: '!room-id',
  };
}

/**
 * Create a matrix vote event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockVote({
  event_id,
  sender = '@user-bob',
  origin_server_ts = 0,
  content = {},
}: {
  event_id?: string;
  sender?: string;
  origin_server_ts?: number;
  content?: Partial<IVote>;
} = {}): RoomEvent<IVote> {
  return {
    type: 'net.nordeck.poll.vote',
    sender,
    content: {
      answerId: '1',
      pollId: 'poll-0',
      'm.relates_to': {
        rel_type: 'm.reference',
        event_id: '$start-event-id',
      },
      ...content,
    },
    origin_server_ts,
    event_id: event_id ?? sender,
    room_id: '!room-id',
  };
}

/**
 * Create a poll settings event with known test data.
 *
 * @remarks Only use for tests
 */
export function mockPollSettings({
  event_id = '$event-id-0',
  content = {},
}: {
  event_id?: string;
  content?: Partial<IPollSettings>;
} = {}): StateEvent<IPollSettings> {
  return {
    type: 'net.nordeck.poll.settings',
    sender: '@user-id',
    content: {
      ...content,
    },
    state_key: '!room-id',
    origin_server_ts: 0,
    event_id,
    room_id: '!room-id',
  };
}
