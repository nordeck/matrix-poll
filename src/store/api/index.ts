/*
 * Copyright 2022 Nordeck IT + Consulting GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { baseApi } from './baseApi';
export {
  pollApi,
  selectAllPolls,
  selectPollById,
  useDeletePollMutation,
  useGetPollsQuery,
  useStartPollMutation,
  useStopPollMutation,
  useUpdatePollMutation,
} from './pollApi';
export {
  selectPollGroup,
  selectPollGroups,
  useDeletePollGroupMutation,
  useGetPollGroupsQuery,
  useUpdatePollGroupMutation,
} from './pollGroupApi';
export {
  useGetPollSettingsQuery,
  usePatchPollSettingsMutation,
} from './pollSettingsApi';
export { powerLevelsApi, useGetPowerLevelsQuery } from './powerLevelsApi';
export {
  roomMemberApi,
  selectActiveRoomMembers,
  selectRoomMember,
  selectRoomMembers,
  useGetRoomMembersQuery,
} from './roomMemberApi';
export { roomNameApi, useGetRoomNameQuery } from './roomNameApi';
export { selectPollsFinished } from './selectPollsFinished';
export { selectPollsOngoing } from './selectPollsOngoing';
export { selectPollsUpcoming } from './selectPollsUpcoming';
export { selectGetVotes, useGetVotes } from './useGetVotes';
export {
  PollInvalidAnswer,
  getVoteAnswerCount,
  selectPollResults,
  usePollResults,
} from './usePollResults';
export type {
  AnswerId,
  GroupResult,
  SelectPollResults,
  Votes,
} from './usePollResults';
export { usePowerLevels } from './usePowerLevels';
export { useRerenderOnPollStatusChange } from './useRerenderOnPollStatusChange';
export { useUserCanVote } from './useUserCanVote';
export { useUserDetails } from './useUserDetails';
export { useVoteMutation, voteApi } from './voteApi';
