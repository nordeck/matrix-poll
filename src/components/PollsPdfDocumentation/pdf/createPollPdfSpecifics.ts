/*
 * Copyright 2022 Nordeck IT + Consulting GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { t } from 'i18next';
import { Content } from 'pdfmake/interfaces';
import { isDefined } from '../../../lib/utils';
import { SelectPollResults } from '../../../store';

export function createPollPdfSpecifics(
  pollResult: SelectPollResults,
  getUserDisplayName: (userId: string) => string,
): Content {
  const pollGroups = pollResult.poll.content.groups;

  const representativeList = pollGroups?.flatMap((g) =>
    Object.entries(g.votingRights)
      .map(([userId, state]) => {
        if (state?.state === 'represented') {
          return {
            group: g.abbreviation,
            representedBy: state.representedBy,
            delegate: userId,
          };
        } else if (state?.state === 'invalid') {
          return {
            group: g.abbreviation,
            delegate: userId,
          };
        }

        return undefined;
      })
      .filter(isDefined),
  );

  if (!pollGroups || !representativeList?.length) {
    return [];
  }

  return [
    {
      style: 'tableExample',
      color: '#444',
      table: {
        widths: ['100%'],
        body: [
          [
            [
              {
                text: t('createPollSpecifics.specifics', 'Specifics'),
                decoration: 'underline',
                margin: [0, 2],
              },
              {
                ul: [
                  ...representativeList.map((r) => ({
                    text: [
                      t(
                        'createPollSpecifics.messageAbsent',
                        'Absent: {{delegate}} ({{group}})',
                        {
                          delegate: getUserDisplayName(r.delegate),
                          group: r.group,
                        },
                      ),
                      ...(r.representedBy
                        ? [
                            { text: ' âž” ', font: 'ZapfDingbats' },
                            t(
                              'createPollSpecifics.messageRepresented',
                              'Represented by: {{delegate}} ({{group}})',
                              {
                                delegate: getUserDisplayName(r.representedBy),
                                group: r.group,
                              },
                            ),
                          ]
                        : []),
                    ],
                    alignment: 'left',
                  })),
                ],
                style: 'list',
              },
            ],
          ],
        ],
      },
    },
  ];
}
