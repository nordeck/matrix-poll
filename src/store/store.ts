/*
 * Copyright 2022 Nordeck IT + Consulting GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { WidgetApi } from '@matrix-widget-toolkit/api';
import { configureStore, Middleware } from '@reduxjs/toolkit';
import { TypedUseSelectorHook, useDispatch, useSelector } from 'react-redux';
import { baseApi } from './api';

const loggerMiddleware: Middleware = () => (next) => (action: unknown) => {
  if (process.env.NODE_ENV === 'development') {
    console.log('LOGGER', action);
  }
  return next(action);
};

export function createStore({
  widgetApi,
  middlewares = [],
}: {
  widgetApi: WidgetApi;
  middlewares?: Middleware[];
}) {
  const store = configureStore({
    reducer: {
      [baseApi.reducerPath]: baseApi.reducer,
    },
    middleware: (getDefaultMiddleware) =>
      getDefaultMiddleware({
        thunk: {
          extraArgument: {
            widgetApi,
          } as ThunkExtraArgument,
        },
      })
        .concat(middlewares)
        .concat(loggerMiddleware)
        .concat(baseApi.middleware),
  });
  return store;
}

/**
 * Extra arguments that are provided to `createAsyncThunk`
 */
export type ThunkExtraArgument = {
  widgetApi: WidgetApi;
};

type StoreType = ReturnType<typeof createStore>;

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<StoreType['getState']>;
export type AppDispatch = StoreType['dispatch'];

export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
