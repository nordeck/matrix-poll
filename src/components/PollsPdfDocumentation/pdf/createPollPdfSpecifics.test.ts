/*
 * Copyright 2022 Nordeck IT + Consulting GmbH
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { mockPoll } from '../../../lib/testUtils';
import { SelectPollResults } from '../../../store';
import { createPollPdfSpecifics } from './createPollPdfSpecifics';

describe('createPollPdfSpecifics', () => {
  const getUserDisplayName = (id: string) => `Name of ${id}`;

  it('should generate a poll specifics content if there is a group and someone is represented', () => {
    const pollResult: SelectPollResults = {
      poll: mockPoll({
        content: {
          groups: [
            {
              id: 'red-party',
              eventId: 'event-id-0',
              abbreviation: 'Red Party',
              color: '#ff0000',
              votingRights: {
                'user-alice': {
                  state: 'active',
                },
                'user-bob': {
                  state: 'represented',
                  representedBy: 'user-backup-1',
                },
              },
            },
            {
              id: 'blue-party',
              eventId: 'event-id-1',
              abbreviation: 'Blue Party',
              color: '#0000ff',
              votingRights: {
                'user-dameon': {
                  state: 'represented',
                  representedBy: 'user-backup-2',
                },
              },
            },
          ],
        },
      }),
      results: {
        votes: {},
      },
      votingRights: [],
    };

    expect(
      createPollPdfSpecifics(pollResult, getUserDisplayName),
    ).toMatchSnapshot();
  });

  it('should generate a poll specifics content if there is a group, someone is representing someone else and someone is not represented', () => {
    const pollResult: SelectPollResults = {
      poll: mockPoll({
        content: {
          groups: [
            {
              id: 'red-party',
              eventId: 'event-id-0',
              abbreviation: 'Red Party',
              color: '#ff0000',
              votingRights: {
                'user-alice': {
                  state: 'active',
                },
                'user-bob': {
                  state: 'represented',
                  representedBy: 'user-backup-1',
                },
                'user-charlie': {
                  state: 'invalid',
                },
              },
            },
            {
              id: 'blue-party',
              eventId: 'event-id-1',
              abbreviation: 'Blue Party',
              color: '#0000ff',
              votingRights: {
                'user-dameon': {
                  state: 'represented',
                  representedBy: 'user-backup-2',
                },
              },
            },
          ],
        },
      }),
      results: {
        votes: {},
      },
      votingRights: [],
    };

    expect(
      createPollPdfSpecifics(pollResult, getUserDisplayName),
    ).toMatchSnapshot();
  });

  it('should not generate a poll specifics content if there is no group', () => {
    const pollResult: SelectPollResults = {
      poll: mockPoll({
        content: {
          groups: undefined,
          startTime: '2020-01-01T03:33:55Z',
        },
      }),
      results: {
        votes: {},
      },
      votingRights: [],
    };

    expect(
      createPollPdfSpecifics(pollResult, getUserDisplayName),
    ).toMatchSnapshot();
  });

  it('should generate a poll specifics content if there is a group with absent delegate without representing', () => {
    const pollResult: SelectPollResults = {
      poll: mockPoll({
        content: {
          groups: [
            {
              id: 'red-party',
              eventId: 'event-id-0',
              abbreviation: 'Red Party',
              color: '#ff0000',
              votingRights: {
                'user-alice': {
                  state: 'active',
                },
                'user-bob': {
                  state: 'invalid',
                },
              },
            },
            {
              id: 'blue-party',
              eventId: 'event-id-1',
              abbreviation: 'Blue Party',
              color: '#0000ff',
              votingRights: {
                'user-dameon': {
                  state: 'active',
                },
              },
            },
          ],
          startTime: '2020-01-01T03:33:55Z',
        },
      }),
      results: {
        votes: {},
      },
      votingRights: [],
    };

    expect(
      createPollPdfSpecifics(pollResult, getUserDisplayName),
    ).toMatchSnapshot();
  });
});
