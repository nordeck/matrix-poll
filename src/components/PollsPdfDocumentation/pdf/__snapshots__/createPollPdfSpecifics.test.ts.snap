// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createPollPdfSpecifics > should generate a poll specifics content if there is a group and someone is represented 1`] = `
[
  {
    "color": "#444",
    "style": "tableExample",
    "table": {
      "body": [
        [
          [
            {
              "decoration": "underline",
              "margin": [
                0,
                2,
              ],
              "text": "Specifics",
            },
            {
              "style": "list",
              "ul": [
                {
                  "alignment": "left",
                  "text": [
                    "Absent: Name of user-bob (Red Party)",
                    {
                      "font": "ZapfDingbats",
                      "text": " ➔ ",
                    },
                    "Represented by: Name of user-backup-1 (Red Party)",
                  ],
                },
                {
                  "alignment": "left",
                  "text": [
                    "Absent: Name of user-dameon (Blue Party)",
                    {
                      "font": "ZapfDingbats",
                      "text": " ➔ ",
                    },
                    "Represented by: Name of user-backup-2 (Blue Party)",
                  ],
                },
              ],
            },
          ],
        ],
      ],
      "widths": [
        "100%",
      ],
    },
  },
]
`;

exports[`createPollPdfSpecifics > should generate a poll specifics content if there is a group with absent delegate without representing 1`] = `
[
  {
    "color": "#444",
    "style": "tableExample",
    "table": {
      "body": [
        [
          [
            {
              "decoration": "underline",
              "margin": [
                0,
                2,
              ],
              "text": "Specifics",
            },
            {
              "style": "list",
              "ul": [
                {
                  "alignment": "left",
                  "text": [
                    "Absent: Name of user-bob (Red Party)",
                  ],
                },
              ],
            },
          ],
        ],
      ],
      "widths": [
        "100%",
      ],
    },
  },
]
`;

exports[`createPollPdfSpecifics > should generate a poll specifics content if there is a group, someone is representing someone else and someone is not represented 1`] = `
[
  {
    "color": "#444",
    "style": "tableExample",
    "table": {
      "body": [
        [
          [
            {
              "decoration": "underline",
              "margin": [
                0,
                2,
              ],
              "text": "Specifics",
            },
            {
              "style": "list",
              "ul": [
                {
                  "alignment": "left",
                  "text": [
                    "Absent: Name of user-bob (Red Party)",
                    {
                      "font": "ZapfDingbats",
                      "text": " ➔ ",
                    },
                    "Represented by: Name of user-backup-1 (Red Party)",
                  ],
                },
                {
                  "alignment": "left",
                  "text": [
                    "Absent: Name of user-charlie (Red Party)",
                  ],
                },
                {
                  "alignment": "left",
                  "text": [
                    "Absent: Name of user-dameon (Blue Party)",
                    {
                      "font": "ZapfDingbats",
                      "text": " ➔ ",
                    },
                    "Represented by: Name of user-backup-2 (Blue Party)",
                  ],
                },
              ],
            },
          ],
        ],
      ],
      "widths": [
        "100%",
      ],
    },
  },
]
`;

exports[`createPollPdfSpecifics > should not generate a poll specifics content if there is no group 1`] = `[]`;
